---
title: "Congruent Transcriptional Responses"
author: "Ben Oyserman, Joris van Steenbrugge, Victoria Pascal Andreu"
date: "2/15/2017"
output:
  html_document: default
  css: custom.css
  toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **An R-package to facilitate high-throughput ecological inferences of time-series RNAseq data.** 
## Introduction:
Micrbiobial Ecologists often use genomic content to infer the biogeochemical processes and interactions within a microbial community. When organisms overlap in their genomic content, they are often inferred to possess similar traits. Conversely, divergent genomic content is indicative of niche differentiation. 

Transcriptomics data may aid in making these ecological inferences by providing information on the regulation of this genetic content. Congruent transcriptional responses (CTRs) of particular metabolic modules in disparate organisms within a community may be indicative of common metabolic features shared by many organisms. Together, organisms with shared metabolic features may contribute to functional redundancy within an community, with important ecological implications.   Conversely, when divergent transcriptional responses are observed across numerous lineages, this is indicative of the disparate niches in the community.

In this package we present a statistical framework to identify CTRs of (pre-defined) metabolic modules in microbial communties. In this manner, genomic bins may be clustered into sub-networks that share responses for particular modules, thereby facilitating inferences on functional redundancy (e.g. intra-cluster), niche partitioning (inter-cluster), and the emergence of complex traits (integrating infromation about multiple modules). 

To demonstrate the utility of this approach, we identify CTRs in polymer storage (polyphosphate, glycogen, and polyhydroxyalkanoates) and amino acid biosynthesis modules of 38 genome bins recovered from a bioreactor operated under conditions that select for organisms capable of storing polymers. We compare different normalization techniques, and look at the role of genomic bin completeness in interpretating results.

![](/Users/benoyserman/Desktop/ctr/CTR_workflow.png)

<center>

Function  | Name              | Description
--------  | ----------------- | -------------  
1         | RNAseq_Normalize  | A function used to normalize raw counts based on various inputs.  
2         | Create_Rank_Columns | Calculate the ranks of each transcript
3         | which_rows_with_no_sd | Identify rows with a SD of 0
4         | Calc_Norm_Euc | Calculates the Euclidean Distance between the normalized ranks
5         | Calc_Jaccard | Calculates the Jaccard Distance between to vectors (Presence/Absence)
6         | Presence_Absence_Matrix | Calculates P/A matrix for all KOs represented >N times in the dataset
7         | Jaccard_Distance_Function | Calculates all pairwise Jaccard Distances for a module
8         | Individual_KOs_Background | Calculates the background distribution for pairwise comparisons of transcripts across bins
9         | Generate_Random_Module | Generates a random module of size N
10        | Background_Distribution_Modules | Calculates a distribution for a random module of size N
11        | NRED_Distance_Function | Calculating normalized rank euclidean distances between modules
12        | P_Distance_Function | Calculating Pearson Correlations between modules
13        | P_NRED_Distance_Function | Calculating composite distances between modules
14        | Cor_Matrix | Output Distance Matrix for both Pearson and Euclidean Distances (implemented in Function F13)
15        | ave_Z_score_Func | Convert array into matrix of average scores (may be used on outputs for F11-F13)
16        | cluster_func | Define clusters using the louvain algorithm
17        | array_fig | Make figure from array

</center>

### A) Importing raw data

  - Structure (Tab deliminated)
    - Locus Tag, Raw Counts, Annotations
  - Import KEGG Orthology Table  
  - Statistics for normalize: Total reads per sample, total reads mapped per genome, edgeR

### B) Calculate background distributions for individual KOs and modules of 'N' KOs

### C) Calculate pairwise distances in the data sets

  - All pairwise Pearson Correlations (PC), Normalized Rank Euclidean Distances (NRED) for each KO<sub>n</sub> in each module, across all bins ($N*\# genomes^2$)
  - Jaccard Distances between Genome A and B ($J_{AB}$)
  - Calculate composite score (CS)  
$$\begin{align*}
CS_{AB}
&=Average(KO_{1},\cdots,KO_{n})/J_{AB}
\end{align*}$$

The number of calculations is equal to $\# genomes^{2* \# KOs}$  

### D) Calculating statistics for each module

The distribution of pairwise distances within each cluster may be compared to a background distribution to calculate a p-value. A small p-value indicates a strong intra-cluster CTR.

In a similar manner, the KOs forming each module (each KO), may be organized based on the distribution of pairwise distances.


# Section A) Importing Raw Data & Data structure

First the necessary librarys are imported, and the random seed is set so that the results are reproducible.

```{r set seet and library upload, results="hide"}
set.seed(1396558)
library("plyr")
library("reshape")
library("igraph")
library('knitr')
library("hexbin")
library("RColorBrewer")
library("ctr")
```

Next, the necessary files are uploaded including

  - 1) the KEGG database 
  - 2) The matrix with transcrtipt abundance and annotations. The format of this matrix is tab delimited: Locus Tag, raw counts, the KO annotation and the Bin.

```{r uploading the matrix}
# First set working directory setwd("/path/to/directory/ctr")
setwd('/path/to/directory/ctr')
KO_pathways_table<-read.table("sample_data/ko00001.keg_tab_delimited",sep="\t",quote = "", row.names = NULL, stringsAsFactors = FALSE)
#KO_modules_table<-read.table("sample_data/KEGG_Modules_14_04_2017_tab_delimited.txt",sep="\t",quote = "", row.names = NULL, stringsAsFactors = FALSE)
KO_modules_table2<-read.table("sample_data/KEGG_Modules_14_04_2017_tab_delimited.txt",sep="\t",quote = "", row.names = NULL, stringsAsFactors = FALSE)

# Barplot of pathway modules
par(mar=c(10,3,3,3))
x <- barplot(sort(table(KO_modules_table2$V1)), xaxt="n")
labs <- names(sort(table(KO_modules_table2$V1)))
text(cex=.5, x=x+1.25, y=-1.25, labs, xpd=TRUE, srt=45, pos=2)

# Barplot of  modules
barplot(table(as.numeric(table(KO_modules_table2$V2))),ylim=c(0,200))

# Load example data set one. Bins obtained from metagenomic assembly -> binning with metabat
RNAseq_Annotated_Matrix_B<-read.table("sample_data/EBPR_RNAseq_Annotation_Matrix",sep="\t",quote = "", row.names = NULL, stringsAsFactors = FALSE, fill = TRUE, col.names=c("Locus_Tag","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","KO"))

# Load example data set one. Bins obtained from metagenomic assembly -> binning with metabat, and a bin replaced with a more complete reference genome 
RNAseq_Annotated_Matrix_BR<-read.table("sample_data/EBPR_RNAseq_Annotation_Matrix_bins_and_reference_sorted",sep=" ",quote = "", row.names = NULL, stringsAsFactors = FALSE, fill = TRUE, col.names=c("Locus_Tag","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","KO"))
```

Modules are defined based on their KEGG Orthology (KO) annotations

```{r manually pre-defined modules, echo = TRUE}
PHA_module <- c("K01895","K00925","K00625","K00626","K00023", "K03821")
polyP_module <- c("K00937","K02040","K02037","K02038","K02036","K07636","K07657","K03306")
Glycogen_module<- c("K00700","K00688","K00703","K01214","K15778","K00975","K00845")
```

In this example, the matrix includes all the bins identified. However, because this method is sensitive to incompleteness we will filter all genomes <80 % complete and with >5% contamination. It is also possible to use reference genomes.

```{r modifying the matrix if necessary, echo = TRUE}
# add a column with the bin number ($Bin). This one-liner parses out the Bin number from the Locus ID
# This is unnecessary if y
RNAseq_Annotated_Matrix_B$Bin<-gsub(".*\\.(.*)\\..*", "\\1", RNAseq_Annotated_Matrix_B[,1])
RNAseq_Annotated_Matrix_BR$Bin<-gsub(".*\\.(.*)\\..*", "\\1", RNAseq_Annotated_Matrix_BR[,1])

# Rename Clade IIA reference to be consistent with bin number
RNAseq_Annotated_Matrix_BR$Bin[grep("CAP2UW1*",RNAseq_Annotated_Matrix_BR$Bin)]<-39

# These are genomes that are  >80% complete <5% contaminated, define a subset matrix with only these genomes
high_quality_bins<-c(8,28,25,7,46,39,22,38,54,53,48,45,31,42,16,33,26,40,36,21,27,17,19,32,14,11,30,43,35,29,23,58,41,20,15,37,49,50)

# Keep only high-quality bins
RNAseq_Annotated_Matrix_B<-RNAseq_Annotated_Matrix_B[which(RNAseq_Annotated_Matrix_B$Bin %in% high_quality_bins),]
RNAseq_Annotated_Matrix_BR<-RNAseq_Annotated_Matrix_BR[which(RNAseq_Annotated_Matrix_BR$Bin %in% high_quality_bins),]
```

If the data is already in the correct format and contains only high-quality bins it should look like this:
```{r Raw Data, include = FALSE}
# Give column names
sample_names<-rep(NA,dim(RNAseq_Annotated_Matrix_B)[2]-3)
for (i in 1:dim(RNAseq_Annotated_Matrix_B)[2]-3) {sample_names[i]<-paste("Sample",i,sep="")}
# colnames(RNAseq_Annotated_Matrix_B)<-c("Locus_ID",sample_names,"KO","Bin")
kable(RNAseq_Annotated_Matrix_B[1:5,], caption = "Raw Data")
```

# Section C) Normalize and calculate background distributions for individual KOs and modules of 'N' KOs

Data will be normalized by the depth of sequencing and the number of reads mapped per genome. It will then be converted to $log_{2}$ scale

### Define inputs
```{r input variables for background distributions, echo=TRUE, results = 'asis'}
matrix_features_B <- new("General_features",
                       high_quality_bins = high_quality_bins,
                       Bin_Column = which(colnames(RNAseq_Annotated_Matrix_B) == "Bin"),
                       sample_names = sample_names,
                       sample_size = length(sample_names),
                       SS = 2,
                       SE = length(sample_names) + 1,
                       RS = which(colnames(RNAseq_Annotated_Matrix_B) == "Bin") + 1,
                       RE = which(colnames(RNAseq_Annotated_Matrix_B) == "Bin") + length(sample_names),
                       no_feature = c(9159700, 4459877, 9826273, 8171512, 9542765, 10522313),
                       ambiguous = c(3940698, 2023389, 4675033, 3308789, 6446272, 5966543),
                       not_aligned = c(0, 19317660, 0, 0, 0, 0),
                       library_size =c(234232896,183166236,228746720,198024002,231567992,259156166)
                       )

matrix_features_BR <- new("General_features",
                       high_quality_bins = high_quality_bins,
                       Bin_Column = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin"),
                       sample_names = sample_names,
                       sample_size = length(sample_names),
                       SS = 2,
                       SE = length(sample_names) + 1,
                       RS = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin") + 1,
                       RE = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin") + length(sample_names),
                       no_feature = c(9159700, 4459877, 9826273, 8171512, 9542765, 10522313),
                       ambiguous = c(3940698, 2023389, 4675033, 3308789, 6446272, 5966543),
                       not_aligned = c(0, 0, 0, 0, 0, 0),
                       library_size = c(234232896,183166236,228746720,198024002,231567992,259156166)
                       )
```

### Normalize
```{r Normalize the raw data, echo=TRUE, results = 'asis'}
# RNAseq_Annotated_Matrix_TMM<-RNAseq_Normalize(RNAseq_Annotated_Matrix,method="RLE")
# RNAseq_Annotated_Matrix_RLE<-RNAseq_Normalize(RNAseq_Annotated_Matrix,method="TMM")
# RNAseq_Annotated_Matrix_RLE_by_bin<-Normalize_by_bin(RNAseq_Annotated_Matrix_RLE)

RNAseq_Annotated_Matrix_B_default<-RNAseq_Normalize(RNAseq_Annotated_Matrix_B, matrix_features_B,
                                                    method = "default")
RNAseq_Annotated_Matrix_B_default_bin<-Normalize_by_bin(RNAseq_Annotated_Matrix_B_default, matrix_features_B)

RNAseq_Annotated_Matrix_BR_default<-RNAseq_Normalize(RNAseq_Annotated_Matrix_BR, matrix_features_BR,
                                                     method = "default")
RNAseq_Annotated_Matrix_BR_default_bin<-Normalize_by_bin(RNAseq_Annotated_Matrix_BR_default, matrix_features_BR)

kable(RNAseq_Annotated_Matrix_B_default[1:5,], caption = "Normalized Data")
kable(RNAseq_Annotated_Matrix_B_default_bin[1:5,], caption = "Normalized Data")

```

### Calculate Ranks

```{r Create Abundance Columns, echo=TRUE, results = 'asis'}
set.seed(1396558) # ties are dealt with randomly, so set.seed must be used to ensure reproducibility
RNAseq_Annotated_Matrix_B_default<-Create_Rank_Columns(RNAseq_Annotated_Matrix_B_default, matrix_features_B)
set.seed(1396558)
RNAseq_Annotated_Matrix_B_default_bin<-Create_Rank_Columns(RNAseq_Annotated_Matrix_B_default_bin, matrix_features_B)

set.seed(1396558) # ties are dealt with randomly, so set.seed must be used to ensure reproducibility
RNAseq_Annotated_Matrix_BR_default<-Create_Rank_Columns(RNAseq_Annotated_Matrix_BR_default, matrix_features_BR)
set.seed(1396558)
RNAseq_Annotated_Matrix_BR_default_bin<-Create_Rank_Columns(RNAseq_Annotated_Matrix_BR_default_bin, matrix_features_BR)



kable(RNAseq_Annotated_Matrix_B_default[1:5,], caption = "Data with Rank Column added")
kable(RNAseq_Annotated_Matrix_B_default_bin[1:5,], caption = "Data with Rank Column added")
```

### How many rows have a SD of zero?
```{r remove rows with standard deviation of 0, echo=TRUE, results = 'asis'}

RNAseq_Annotation_Matrix_no_sd_of_zero_B<-which_rows_with_no_sd(RNAseq_Annotated_Matrix_B_default_bin, 
                                                                matrix_features_B)
RNAseq_Annotation_Matrix_no_sd_of_zero_BR<-which_rows_with_no_sd(RNAseq_Annotated_Matrix_BR_default_bin,
                                                                 matrix_features_BR)

```

Of `r dim(RNAseq_Annotated_Matrix)[1]` rows (transcripts), `r dim(RNAseq_Annotated_Matrix)[1]-dim(RNAseq_Annotation_Matrix_no_sd_of_zero)[1]` had a standard deviation of 0. These are not included when calculating background distibutions.

Here we calculate the background distribution of the individual KOs, and test whether there is a significant difference between a random pairing of genes in two bins, and a random pairing of genes *with same function* in two bins. When multiple genes with a given function are present in a genome, two alternative strategies are taken whereby either 1) a random pairwise distance is used or 2) the minimum distance is used. 

```{r, Calculates background distributions for individual KOs, echo=TRUE, results = 'asis' }

I_KOs_Background_B <- Individual_KOs_Background(RNAseq_Annotation_Matrix_no_sd_of_zero_B,matrix_features_B,10000)
I_KOs_Background_BR <- Individual_KOs_Background(RNAseq_Annotation_Matrix_no_sd_of_zero_BR,matrix_features_BR,10000)

# What would a background distribution look like if there were no ties (e.g. genes with the same counts)?
# random_euclidean_distances<-rep(NA,10000)
# for (i in 1:10000) {random_euclidean_distances[i]<-Calc_Norm_Euc(sample(seq(.001,1,.001),6),sample(seq(.001,1,.001),6))}

```

Based on these background statistics, means & standard deviations for calculating Z scores may be calcualted. To calculate the Z score, here we used the mu and sd from the random background distribution and account for multiple comparisons when a KO is present in multiple copies in a genome.

```{r}
Z_scores_B<-calc_Z_scores(I_KOs_Background_B)
Z_scores_BR<-calc_Z_scores(I_KOs_Background_BR)
```

These Z-scores may be combined into a composite score. Below is a visualization of Z scores from the random background distributions (accounting for multiple comparisons). It is clear that when genes share a KO, they also share an expression profile.

```{r, some nice plots for the background distributions echo=TRUE}
rf <- colorRampPalette(rev(brewer.pal(11,'Spectral')))

plot(hexbin(Z_scores_B$Z_H_random), colramp=rf, mincnt=1,maxcnt=max(hexbin(Z_scores_B$Z_H_KO)@count),ylab="NRED",xlab="PC")
plot(hexbin(Z_scores_B$Z_H_KO), colramp=rf, mincnt=1, maxcnt=max(hexbin(Z_scores_B$Z_H_KO)@count),ylab="NRED",xlab="PC")

plot(hexbin(Z_scores_BR$Z_H_random), colramp=rf, mincnt=1, maxcnt=max(hexbin(Z_scores_BR$Z_H_KO)@count),ylab="NRED",xlab="PC")
plot(hexbin(Z_scores_BR$Z_H_KO), colramp=rf, mincnt=1, maxcnt=max(hexbin(Z_scores_BR$Z_H_KO)@count),ylab="NRED",xlab="PC")


t_test_KO_random_pearson_B<-t.test(I_KOs_Background_B$KO_pairwise_gene_pearson,I_KOs_Background_B$random_pairwise_gene_pearson, alternative="greater") # x > y (NULL)
t_test_KO_random_euclidean_B<-t.test(I_KOs_Background_B$random_pairwise_gene_euclidean,I_KOs_Background_B$H_KO_pairwise_gene_euclidean, alternative="greater") # x > y (NULL)

t_test_KO_random_pearson_BR<-t.test(I_KOs_Background_BR$KO_pairwise_gene_pearson,I_KOs_Background_BR$random_pairwise_gene_pearson, alternative="greater") # x > y (NULL)
t_test_KO_random_euclidean_BR<-t.test(I_KOs_Background_BR$random_pairwise_gene_euclidean,I_KOs_Background_BR$H_KO_pairwise_gene_euclidean, alternative="greater") # x > y (NULL)

par(mfrow=c(2,2),mar=c(3,3,3,1))
# plot 1
plot(density(I_KOs_Background_B$random_pairwise_gene_pearson,adjust = 2,na.rm=TRUE),ylim=c(0,1),xlab="",ylab="",main="")
points(density(I_KOs_Background_B$KO_pairwise_gene_pearson,adjust = 2),typ="l",col="blue")
mtext(paste("p-value = ",signif(t_test_KO_random_pearson_B$p.value,2)),side=3,col="blue",padj=2,cex=.75)
title(ylab="Density", line=2, cex.lab=1)
title(xlab="PC", line=2, cex.lab=1)

# plot 2
plot(density(I_KOs_Background_BR$H_random_pairwise_gene_pearson,adjust = 2),ylim=c(0,1),xlab="",ylab="",main=" ")
points(density(I_KOs_Background_BR$H_KO_pairwise_gene_pearson,adjust = 2),typ="l",col="red")
mtext(paste("p-value = ",signif(t_test_KO_random_pearson_BR$p.value,2)),side=3,col="red",padj=2,cex=.75)
title(ylab="Density", line=2, cex.lab=1)
title(xlab="PC", line=2, cex.lab=1)

# plot 3
plot(density(I_KOs_Background_B$random_pairwise_gene_euclidean,adjust = 2),typ="l" ,ylim=c(0,1.25),xlab="",ylab="",main="")
points(density(I_KOs_Background_B$KO_pairwise_gene_euclidean,adjust = 2),typ="l",col="blue")
title(ylab="Density", line=2, cex.lab=1)
title(xlab="NRED", line=2, cex.lab=1)
mtext(paste("p-value = ",signif(t_test_KO_random_euclidean_B$p.value,2)),side=3,col="blue",padj=2,cex=.75)

# plot 4
plot(density(I_KOs_Background_BR$H_random_pairwise_gene_euclidean,adjust = 2),typ="l" ,ylim=c(0,1.25),xlab="",ylab="",main="")
points(density(I_KOs_Background_BR$H_KO_pairwise_gene_euclidean,adjust = 2),typ="l",col="red")
title(ylab="Density", line=2, cex.lab=1)
title(xlab="NRED", line=2, cex.lab=1)
title(" \n\nComparison of random & functional \n pairwise comparisons", outer=TRUE) 
mtext(paste("p-value = ",signif(t_test_KO_random_euclidean_BR$p.value,2)),side=3,col="red",padj=2,cex=.75)
```

# Section D) Calculate background distributions for modules of 'N' KOs

```{r Backround Distributions for Modules, echo=FALSE}   
# RNAseq_Annotated_Matrix_BR_default_by_bin
# RNAseq_Annotated_Matrix_default_by_bin

ptm <- proc.time()
for (i in 1:1000) {
!length(table(test_matrix[4,]))==1
}
proc.time() - ptm

ptm <- proc.time()
for (i in 1:1000) {
!sd(test_matrix[4,])==0
}
proc.time() - ptm

ptm <- proc.time()
Random_Background_Module_Distances_6_BR<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_BR_default_bin, matrix_features_BR, Z_scores_BR, 6, 1000)
proc.time() - ptm

Random_Background_Module_Distances_7_BR<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_BR_default_bin, matrix_features_BR, Z_scores_BR, 7, 1000)
Random_Background_Module_Distances_8_BR<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_BR_default_bin, matrix_features_BR, Z_scores_BR, 8, 1000)

Random_Background_Module_Distances_6_B<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_B_default_bin, matrix_features_B, Z_scores_B, 6, 1000)
Random_Background_Module_Distances_7_B<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_B_default_bin, matrix_features_B, Z_scores_B, 7, 1000)
Random_Background_Module_Distances_8_B<-Background_Distribution_Modules(RNAseq_Annotated_Matrix_B_default_bin, matrix_features_B, Z_scores_B, 8, 1000)
proc.time() - ptm

par(mfrow=c(1,2))
plot(density(Random_Background_Module_Distances_6_B,na.rm=TRUE),col="red",main="Background distributions of \nmodules of size \"N\"\nB",ylim=c(0,1),xlim=c(-4,4),cex.main=.75)
points(density(Random_Background_Module_Distances_7_B,na.rm=TRUE),main="N=7",col="blue",type="l")
points(density(Random_Background_Module_Distances_8_B,na.rm=TRUE),main="N=8",col="green",type="l")
legend("topright",legend=c("N=6","N=7","N=8"),col=c("red","blue","green"),lty=c(1,1,1))

plot(density(Random_Background_Module_Distances_6_BR,na.rm=TRUE),col="red",main="Background distributions of modules \nof size \"N\"\nBR",ylim=c(0,1),xlim=c(-4,4),cex.main=.75)
points(density(Random_Background_Module_Distances_7_BR,na.rm=TRUE),main="N=7",col="blue",type="l")
points(density(Random_Background_Module_Distances_8_BR,na.rm=TRUE),main="N=8",col="green",type="l")
legend("topright",legend=c("N=6","N=7","N=8"),col=c("red","blue","green"),lty=c(1,1,1))
```


## Section 5) Calculating statsitics for each module

#### Statistics for the PHA module
```{r plotting the groups PHA, echo=FALSE}

# Calculating distances based on univariate (Normalized Rank Euclidean Distances OR Pearson Correlation) or multivariate (Statistically)
# PHA_module_NRED <- NRED_Distance_Function(PHA_module) # need to change these codes to have RNAseq_Annotated_Matrix as a local rather than global variable
# PHA_module_P <- P_Distance_Function(PHA_module) # need to change these codes to have RNAseq_Annotated_Matrix as a local rather than global variable

# calculate composite scores and jaccard distances, first for B
PHA_module_P_NRED_B <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,
                                                Z_scores_B, PHA_module)
Jaccard_Distance_PHA_B <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,PHA_module)
PHA_clustering_results_P_NRED_B <-cluster_func(PHA_module_P_NRED_B$combined,Jaccard_Distance_PHA_B)

Glycogen_module_P_NRED_B <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,
                                                     Z_scores_B, Glycogen_module)
Jaccard_Distance_Glycogen_B <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,Glycogen_module)
Glycogen_clustering_results_P_NRED_B <-cluster_func(Glycogen_module_P_NRED_B$combined,Jaccard_Distance_Glycogen_B)

polyP_module_P_NRED_B <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,
                                                  Z_scores_B, polyP_module)
Jaccard_Distance_polyP_B <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_B_default_bin,polyP_module)
polyP_clustering_results_P_NRED_B <-cluster_func(polyP_module_P_NRED_B$combined,Jaccard_Distance_polyP_B)

# than for BR
PHA_module_P_NRED_BR <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,
                                                 Z_scores_B, PHA_module)
Jaccard_Distance_PHA_BR <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,PHA_module)
PHA_clustering_results_P_NRED_BR <-cluster_func(PHA_module_P_NRED_BR$combined,Jaccard_Distance_PHA_BR)

Glycogen_module_P_NRED_BR <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,
                                                      Z_scores_B, Glycogen_module)
Jaccard_Distance_Glycogen_BR <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,Glycogen_module)
Glycogen_clustering_results_P_NRED_BR <-cluster_func(Glycogen_module_P_NRED_BR$combined,Jaccard_Distance_Glycogen_BR)

polyP_module_P_NRED_BR <- P_NRED_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,
                                                   Z_scores_B, polyP_module)
Jaccard_Distance_polyP_BR <- Jaccard_Distance_Function(RNAseq_Annotated_Matrix_BR_default_bin,polyP_module)
polyP_clustering_results_P_NRED_BR <-cluster_func(polyP_module_P_NRED_BR$combined,Jaccard_Distance_polyP_BR)

# replace NAs with a large distance
# Glycogen_pvalues_BR$distances[[2]][which(upper.tri(Glycogen_pvalues_BR$distances[[2]], diag = FALSE))[is.na(Glycogen_pvalues_BR$distances[[2]][upper.tri(Glycogen_pvalues_BR$distances[[2]], diag = FALSE)])]]<-4

# Prune clusters #
PHA_pvalues_BR<-ClusterPrune(PHA_clustering_results_P_NRED_BR,Random_Background_Module_Distances_6_BR)
Glycogen_pvalues_BR<-ClusterPrune(Glycogen_clustering_results_P_NRED_BR,Random_Background_Module_Distances_7_BR)
polyP_pvalues_BR<-ClusterPrune(polyP_clustering_results_P_NRED_BR,Random_Background_Module_Distances_8_BR)
Leucine_pvalues_BR<-ClusterPrune(Leucine_clustering_results_P_NRED_BR,Random_Background_Module_Distances_6_BR)

PHA_pvalues_B<-ClusterPrune(PHA_clustering_results_P_NRED_B,Random_Background_Module_Distances_6_B)
Glycogen_pvalues_B<-ClusterPrune(Glycogen_clustering_results_P_NRED_B,Random_Background_Module_Distances_7_B)
polyP_pvalues_B<-ClusterPrune(polyP_clustering_results_P_NRED_B,Random_Background_Module_Distances_8_B)


plot(hclust(as.dist(t(as.matrix(PHA_pvalues_BR$distances[[3]])))))
```
