---
title: "Congruent Transcriptional Responses"
author: "Ben Oyserman, Joris van Steenbrugge, Victoria Pascal Andreu"
date: "2/15/2017"
output:
  html_document: default
  css: custom.css
  toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **An R-package to facilitate high-throughput ecological inferences of time-series RNAseq data.** 
## Introduction:
Micrbiobial Ecologists often use genomic content to infer the biogeochemical processes and interactions within a microbial community. When organisms overlap in their genomic content, they are often inferred to possess similar traits. Conversely, divergent genomic content is indicative of niche differentiation. 

Transcriptomics data may aid in making these ecological inferences by providing information on the regulation of this genetic content. Congruent transcriptional responses (CTRs) of particular metabolic modules in disparate organisms within a community may be indicative of common metabolic features shared by many organisms. Together, organisms with shared metabolic features may contribute to functional redundancy within an community, with important ecological implications.   Conversely, when divergent transcriptional responses are observed across numerous lineages, this is indicative of the disparate niches in the community.

In this package we present a statistical framework to identify CTRs of (pre-defined) metabolic modules in microbial communties. In this manner, genomic bins may be clustered into sub-networks that share responses for particular modules, thereby facilitating inferences on functional redundancy (e.g. intra-cluster), niche partitioning (inter-cluster), and the emergence of complex traits (integrating infromation about multiple modules). 

To demonstrate the utility of this approach, we identify CTRs in polymer storage (polyphosphate, glycogen, and polyhydroxyalkanoates) and amino acid biosynthesis modules of 38 genome bins recovered from a bioreactor operated under conditions that select for organisms capable of storing polymers.

![](/home/joris/tools/ctr/CTR_workflow.png)


# Section A) Importing Raw Data & Data structure

First the necessary librarys are imported.

```{r set seet and library upload, results='hide', message=FALSE}
library("reshape")
library("knitr")
library("ctr")
library("RColorBrewer")
library('igraph')
library('stringr')
library('RCy3')
```

Next, the necessary files are uploaded including

  - 1) the KEGG database 
  - 2) The matrix with transcrtipt abundance and annotations. The format of this matrix is tab delimited: Locus Tag, raw counts, the KO annotation and the Bin.

```{r uploading the matrix}
# First set working directory setwd("/path/to/directory/ctr")
 setwd('/home/joris/tools/ctr')
# sourceDirectory("/path/to/directory/ctr/R/");
KO_pathways_table <- read.table("sample_data/ko00001.keg_tab_delimited",sep = "\t",
                              quote = "", row.names = NULL, stringsAsFactors = FALSE)

KO_modules_table2 <- read.table("sample_data/KEGG_Modules_14_04_2017_tab_delimited.txt",
                              sep = "\t",quote = "", row.names = NULL, 
                              stringsAsFactors = FALSE)


# Load data set, the bins in the sample dataset are obtained from metagenomic assembly -> binning with metabat, and a bin replaced with a more complete reference genome.

# Creates a vector with Sample1, Sample2, .., SampleN)
# where N is the number of samples
N = 6 # in this case N = 6
sample_names <- sapply(1:N, function(i) paste("Sample", i, sep=""))
RNAseq_Annotated_Matrix_BR <- read.table("sample_data/EBPR_RNAseq_Annotation_Matrix_bins_and_reference_sorted",
                                       sep = " ", quote = "", row.names = NULL, 
                                       stringsAsFactors = FALSE, fill = TRUE, 
                                       col.names = c("Locus_Tag", sample_names , "KO"))
```

In this example, the matrix includes all the bins identified. However, because this method is sensitive to incompleteness we will filter all genomes < 80 % complete and with > 5% contamination. It is also possible to use reference genomes.

```{r modifying the matrix if necessary, echo = TRUE}
# add a column with the bin number ($Bin). This one-liner parses out the Bin number from the Locus ID
# This is unnecessary if y
RNAseq_Annotated_Matrix_BR$Bin <- gsub(".*\\.(.*)\\..*", "\\1", 
                                       RNAseq_Annotated_Matrix_BR[, 1])

# Rename Clade IIA reference to be consistent with bin number
RNAseq_Annotated_Matrix_BR$Bin[grep("CAP2UW1*", 
                                    RNAseq_Annotated_Matrix_BR$Bin)] <- 39

# These are genomes that are  >80% complete <5% contaminated, define a subset matrix with only these genomes
high_quality_bins <- c(8, 28, 25, 7, 46, 39, 22, 38, 54, 53, 48, 45, 31, 42, 16, 
                       33, 26, 40, 36, 21, 27, 17, 19, 32, 14, 11, 30, 43, 35, 
                       29, 23, 58, 41, 20, 15, 37, 49, 50)

# Keep only high-quality bins
RNAseq_Annotated_Matrix_BR <- RNAseq_Annotated_Matrix_BR[which(
                                RNAseq_Annotated_Matrix_BR$Bin %in% high_quality_bins),]

sample_names <- sapply(seq((dim(RNAseq_Annotated_Matrix_BR)[2] - 3)), 
                       function(x) return(paste("Sample", x, sep="")))
```




### Define meta-data
```{r input variables for background distributions, echo=TRUE, results = 'asis'}
matrix_features_BR <- new("General_features",
               high_quality_bins = high_quality_bins,
               Bin_Column = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin"),
               sample_names = sample_names,
               sample_size = length(sample_names),
               SS = 2,
               SE = length(sample_names) + 1,
               RS = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin") + 1,
               RE = which(colnames(RNAseq_Annotated_Matrix_BR) == "Bin") + length(sample_names),
               no_feature = c(9159700, 4459877, 9826273, 8171512, 9542765, 10522313), # This must be updated!
               ambiguous = c(3940698, 2023389, 4675033, 3308789, 6446272, 5966543), # This must be updated!
               not_aligned = c(0, 0, 0, 0, 0, 0),
               library_size = c(234232896, 183166236, 228746720, 198024002, 231567992, 259156166),
               Pairwise_Bin_Array_Presence = Presence_Absence_Matrix(RNAseq_Annotated_Matrix_BR),
               no_annotation = which(names((table(RNAseq_Annotated_Matrix_BR$KO))) == ""),
               All_KOs = names(table(RNAseq_Annotated_Matrix_BR$KO))[
                 -which(names((table(RNAseq_Annotated_Matrix_BR$KO))) == "")]
              )

biased_modules <- c("M00083","M00572","M00307","M00009","M00011","M00532","M00036",
                    "M00375","M00374","M00373","M00088","M00095","M00086","M00254",
                    "M00237","M00258","M00239")

list_of_all_modules <- make_module_lists(KO_modules_table2)
biased_idx <- which(names(list_of_all_modules) %in% biased_modules)
list_of_all_modules <- list_of_all_modules[-biased_idx]
list_of_all_modules_KOs <- unique(unlist(list_of_all_modules, use.names=FALSE))

sublist_of_all_modules <-  make_module_lists(KO_modules_table2)[1:10]
sublist_of_all_modules_KOs <- unique(unlist(sublist_of_all_modules, use.names=FALSE))
plot(table(lengths(list_of_all_modules)))


polymers_KOs <- unique(unlist(polymers, use.names=FALSE))
```

### Minimal Usage
Different data processing steps such as normalization and the random background generation
of Individual KO terms and of modules will be calculated to do further analysis with.

This step partially is able to run in parallel on multiple processing (CPU) cores. In general more cores means shorter running times.
The parallel_cores parameter in the transcriptional_responses$Run() function can be set to the number of cores you want to use.

If you are not sure how many CPU cores your machine has:  
On Linux:    
  Enter in a terminal: `nproc`   
On Mac:    
  Enter in a terminal: `sysctl hw.ncpu`

```{r message=FALSE, results='hide'}
transcriptional_responses <- CTR$new(RNAseq_Annotated_Matrix = RNAseq_Annotated_Matrix_BR,
                                     matrix_features = matrix_features_BR)
#Run all analyses
transcriptional_responses$Run(iterations = 10000,
                              random_module_sizes = 1:20,
                              parallel_cores = 3)

transcriptional_responses$AssociationMatrix(polymers_KOs, polymers)

```

### Random Background for individual KO terms
```{r}
transcriptional_responses$plotIndividualBackgroundDist()
``` 

### Random Background for KO modules
```{r}
transcriptional_responses$plotModuleBackgroundDist()


colors <- c("red", "green", "yellow", "blue", "purple", "black", "orange", "grey", 
            "pink", "cyan")
bg_distance_names <- c("2", "3","4","5", "6","7", "8", "9","10")
legend <- sapply(bg_distance_names, function(x) paste('N = ', x, sep=""))

plot(density(transcriptional_responses$bg_distance_modules[[1]],na.rm=TRUE),
     ylim = c(0, 1), xlim = c(-4, 4), cex.main = .75, col = colors[1],
     main = "Background distributions of module sizes 2 - 10")
points(density(transcriptional_responses$bg_distance_modules[[2]], na.rm=TRUE), 
       type="l", col = colors[2])
points(density(transcriptional_responses$bg_distance_modules[[3]], na.rm=TRUE), 
       type="l", col = colors[3])
points(density(transcriptional_responses$bg_distance_modules[[4]], na.rm=TRUE), 
       type="l", col = colors[4])
points(density(transcriptional_responses$bg_distance_modules[[5]], na.rm=TRUE), 
       type="l", col = colors[5])
points(density(transcriptional_responses$bg_distance_modules[[6]], na.rm=TRUE), 
       type="l", col = colors[6])
points(density(transcriptional_responses$bg_distance_modules[[7]], na.rm=TRUE), 
       type="l", col = colors[7])
points(density(transcriptional_responses$bg_distance_modules[[8]], na.rm=TRUE), 
       type="l", col = colors[8])
points(density(transcriptional_responses$bg_distance_modules[[9]], na.rm=TRUE), 
       type="l", col = colors[9])

legend("topleft",legend = legend, col = colours, pch = 1)
```


### Association rule mining
In the previous steps the variable `All_association_matrix` was created.
All_association_matrix contains information of what genomes/bins have a certain KO module activated and what transcriptional profile. This information is used to determine association rules.
```{r, eval=FALSE}
#filtered_association_matrix <- transcriptional_responses$All_association_matrix[, as.numeric(which(unlist(lapply(list_of_all_modules, length))<11))]
association_rules <-  Apriori$new(dataset = transcriptional_responses$All_association_matrix)
association_rules$run_apriori(supp = 0.4, conf = 0.9 )

```

### Plot association rules in cytoscape
This feature is currently experimental. It can only work with Cytoscape version 3.4 or higher running on Java 8 (Cytoscape must be up and running before running the line below). Inside Cytoscape the `cyRest` app has to be installed (version 3.3 or higher).  
`association_rules$rules_dataframe` corresponds to the rule set generated by `Apriori` in an R dataframe structure (calculated in the previous step).  
`KO_modules_table2` is declared at the beginning of this document and contains the corresponding KO module for each KO term.
`N` is the amount of rules (sorted on support or lift) to plot.
```{r, eval=FALSE}
Draw_Network(association_rules$rules_dataframe, KO_modules_table2, N = 400)
```
